src/session_impl.cpp

```diff
@@ -283,6 +278,72 @@ namespace aux {
 		}
 	}
 
+	void expand_devices(span<ip_interface const> const ifs
+		, span<ip_route const> const routes
+		, std::vector<listen_endpoint_t>& eps)
+	{
+		for (auto& ep : eps)
+		{
+			auto const iface = ep.device.empty()
+				? std::find_if(ifs.begin(), ifs.end(), [&](ip_interface const& ipface)
+					{
+						return match_addr_mask(ipface.interface_address, ep.addr, ipface.netmask);
+					})
+				: std::find_if(ifs.begin(), ifs.end(), [&](ip_interface const& ipface)
+					{
+						return ipface.name == ep.device
+							&& match_addr_mask(ipface.interface_address, ep.addr, ipface.netmask);
+					});
+
+			if (iface == ifs.end())
+			{
+				// we can't find which device this is for, just assume we can't
+				// reach anything on it
+				ep.netmask = build_netmask(0, ep.addr.is_v4() ? AF_INET : AF_INET6);
+				continue;
+			}
+
+			ep.netmask = iface->netmask;
+
+			bool const v4 = iface->interface_address.is_v4();
+
+			// also record whether the device has a gateway associated with it
+			// (which indicates it can be used to reach the internet)
+			// only gateways inside the interface's network count
+			bool const has_gateway =
+				(v4 || !iface->interface_address.to_v6().is_link_local())
```

Comment:
Use `is_local` instead of `is_link_local`, link-local is just one of several local address ranges.

Reply:
the more I think about this, it's not obvious to me that this actually matters. with the check that the gateway is inside the network specified in the `ip_interface`, I don't think it will be mis-associated with a local interface. It would only happen if the gateway's IP address also falls within the local IP range. like, `127.1.1.1` or something. Which would be weird, but would it necessarily be wrong?
`is_local()` is pretty broad and covers IPv4 as well, and I think a lot of the cases it covers wouldn't be necessary.

I'm not very familiar with the semantics of scope IDs in IPv6. But is it possible to have an IP address that looks global but is local just by the scope ID?

Reply:
actually, `is_local()` checks `is_site_local()`. I'm not an expert on this obviously, but I would imagine you can have a local "site" that consists of multiple subnets, where you actually need a gateway between them.

In the current use of `is_local()`this makes sense (I think), since it's used to determine whether to apply rate limits or not (at least it did historically), I suppose this is mostly based on IP filters now.

In fact, I get the impression that there shouldn't be a special case here. I think it's possible, and perhaps even likely, that some sites run a private IPv6 network, with all nodes in the `fc00::/7` block (i.e. local addresses), but still have this private network connected to the internet, via NAT. In that case, you would have an interface with a local IPv6 address, and still have a gateway to the internet. As long as that gateway also live within the same private network, and is reachable from this interface, that seems like a legitimate set up.

Reply:
What you describe is NAT66 which is rare and considered bad practice. Even if we wanted to support it, which I oppose, we'd have to actively probe each local address to see if it can reach outside the local network before flagging it with `has_gateway`. Otherwise in the common case we'd constantly be trying to send global traffic on local addresses which cannot reach the internet.

Reply:
Site-local addresses are specified as non-routable, so they should not be used across subnets. Anyways they were deprecated a long time ago (RFC 3879) so I doubt they're actually used today.